cli/util/uploader.go

```diff
@@ -0,0 +1,352 @@
+	}
+	ctx.App.Writer.Write([]byte(fmt.Sprintf("Chain block height: %d\n", currentBlockHeight)))
+
+	maxBlockIndex, err := fetchMaxBlockIndex(ctx.Context, clientSDK, containerID, acc.PrivateKey(), uint(currentBlockHeight), attribute)
+	if err != nil {
+		return fmt.Errorf("failed to fetch max block index from container: %w", err)
+	}
+	ctx.App.Writer.Write([]byte(fmt.Sprintf("Max block index in NeoFS container: %d\n", maxBlockIndex)))
+	if maxBlockIndex == 0 {
+		maxBlockIndex = -1
+	}
+
+	if maxBlockIndex > int(currentBlockHeight) {
+		return fmt.Errorf("no new blocks to upload. Max index in NeoFS: %d, current height: %d", maxBlockIndex, currentBlockHeight)
+	}
+
+	for blockIndex := maxBlockIndex + 1; blockIndex <= int(currentBlockHeight); blockIndex++ {
+		blockData, err := fetchBlockData(rpc, uint(blockIndex))
+		if err != nil {
+			return fmt.Errorf("failed to fetch block %d: %w", blockIndex, err)
+		}
+
+		err = uploadBObjWithSlicer(ctx.Context, clientSDK, *acc, containerID, blockData, attribute, strconv.Itoa(blockIndex), sessionToken)
+		if err != nil {
+			return fmt.Errorf("failed to upload block %d: %w", blockIndex, err)
+		}
+
+		if blockIndex%1000 == 0 {
+			ctx.App.Writer.Write([]byte(fmt.Sprintf("Successfully uploaded block: %d", blockIndex)))
+		}
+	}
+
+	err = updateIndexFiles(ctx, clientSDK, containerID, *acc, uint(currentBlockHeight), indexFileAttribute, attribute)
+	if err != nil {
+		return fmt.Errorf("failed to update index files after upload: %w", err)
+	}
+	ctx.App.Writer.Write([]byte("Upload completed successfully.\n"))
+	return nil
+}
+
+// fetchMaxBlockIndex searches for the maximum block index in the container.
+func fetchMaxBlockIndex(ctx context.Context, clientSDK *client.Client, containerID cid.ID, priv *keys.PrivateKey, currentHeight uint, attributeKey string) (int, error) {
+	height := int(currentHeight)
+	var (
+		finalResult int
+		finalErr    error
+	)
+
+	searchBatch := func(start, end int) (int, bool, error) {
+		prm := client.PrmObjectSearch{}
+		filters := object.NewSearchFilters()
+		filters.AddFilter(attributeKey, fmt.Sprintf("%d", start), object.MatchNumGE)
+		filters.AddFilter(attributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
+		prm.SetFilters(filters)
+
+		objectIDs, err := neofs.ObjectSearch(ctx, clientSDK, priv, containerID.String(), prm)
+		if err != nil {
+			return 0, false, fmt.Errorf("failed to search objects from %d to %d: %w", start, end, err)
+		}
+
+		numOIDs := len(objectIDs)
+		fmt.Printf("Found %d blocks between %d and %d", numOIDs, start, end)
+
+		if numOIDs == 0 {
+			return 0, false, nil // Keep searching.
+		}
+
+		// Return the max block index found in this batch
+		maxInBatch := start + numOIDs - 1
+
+		// Stop immediately after finding the first non-empty batch
+		return maxInBatch, true, nil
+	}
+
+	for ; height > 0; height -= searchBatchSize {
+		startIndex := height - searchBatchSize + 1
+		if startIndex < 0 {
+			startIndex = 0
+		}
```

Comment:
We need to check if there are any bugs in this code, because the following logs seem to be caused by a single script run:
```
[sergio@kaede:~]% journalctl -f -u neogo-block-fetcher-fschain-testnet.service
Oct 02 10:41:57 kaede neogo-block-fetcher[737354]: 2024/10/02 10:41:57 Found 0 blocks between 2884789 and 2894788
Oct 02 10:42:13 kaede neogo-block-fetcher[737354]: 2024/10/02 10:42:13 Found 0 blocks between 2874789 and 2884788
Oct 02 10:42:29 kaede neogo-block-fetcher[737354]: 2024/10/02 10:42:29 Found 0 blocks between 2864789 and 2874788
Oct 02 10:42:44 kaede neogo-block-fetcher[737354]: 2024/10/02 10:42:44 Found 0 blocks between 2854789 and 2864788
Oct 02 10:43:00 kaede neogo-block-fetcher[737354]: 2024/10/02 10:43:00 Found 0 blocks between 2844789 and 2854788
Oct 02 10:43:19 kaede neogo-block-fetcher[737354]: 2024/10/02 10:43:19 Found 0 blocks between 2834789 and 2844788
Oct 02 10:43:40 kaede neogo-block-fetcher[737354]: 2024/10/02 10:43:40 Found 0 blocks between 2824789 and 2834788
Oct 02 10:43:56 kaede neogo-block-fetcher[737354]: 2024/10/02 10:43:56 Found 0 blocks between 2814789 and 2824788
Oct 02 10:44:12 kaede neogo-block-fetcher[737354]: 2024/10/02 10:44:12 Found 0 blocks between 2804789 and 2814788
Oct 02 10:44:28 kaede neogo-block-fetcher[737354]: 2024/10/02 10:44:28 Found 0 blocks between 2794789 and 2804788
Oct 02 10:44:43 kaede neogo-block-fetcher[737354]: 2024/10/02 10:44:43 Found 0 blocks between 2784789 and 2794788
Oct 02 10:44:59 kaede neogo-block-fetcher[737354]: 2024/10/02 10:44:59 Found 0 blocks between 2774789 and 2784788
```

Reply:
OK, not relevant.